Race condition in approle login with identical role-id

<!-- Please reserve GitHub issues for bug reports and feature requests.

For questions, the best place to get answers is on our [mailing list](https://groups.google.com/forum/#!forum/vault-tool), as they will get more visibility from experienced users than the issue tracker.

Please note: We take Vault's security and our users' trust very seriously. If you believe you have found a security issue in Vault, please responsibly disclose by contacting us at security@hashicorp.com. Our PGP key is available at [our security page](https://www.hashicorp.com/security/).
-->

**Environment:**
<!-- The version can be retrieved with `vault version`. -->

* Vault Version: `Vault v0.9.0 ('bdac1854478538052ba5b7ec9a9ec688d35a3335')`
* Operating System/Architecture: OSX 10.12.6

**Vault Config File:**

```
storage "file" {
  path = "/tmp/vault/data"
}

listener "tcp" {
  address     = "127.0.0.1:8200"
  tls_disable = 1
}
```

**Startup Log Output:**

```
...
2018/01/02 17:01:10.045995 [INFO ] core: post-unseal setup complete
2018/01/02 17:01:10.149368 [INFO ] core: enabled credential backend: path=approle/ type=approle
2018/01/02 17:01:10.742978 [DEBUG] core: creating a new entity: alias=&{approle auth_approle_d638ab40 7abd8541-d9bc-3936-15ad-964d8217732d}
2018/01/02 17:01:10.743025 [DEBUG] core: creating a new entity: alias=&{approle auth_approle_d638ab40 7abd8541-d9bc-3936-15ad-964d8217732d}
==> Vault shutdown triggered
2018/01/02 17:01:12.301662 [DEBUG] core: marked as sealed
...
```

**Expected Behavior:**
Approle `/auth/approle/login` should always succeed given a valid role-id and secret-id.

**Actual Behavior:**
Login returns 500 response with error message like `alias "6701e7da-142f-12a9-22cf-ebf5b343414d" in already tied to a different entity "a29c7397-516b-2595-8c36-ff8ca0af7edd"`.

**Steps to Reproduce:**

Perform multiple logins to an approle within μsec from one another. I am unable to estimate threshold at which this race condition is encountered (could be between 1 and 1000 μsec).

Logging from my client implementation (all values below are generated for issue reporting purpose):

```
...
2018-01-02 17:09:47,559 [DEBUG] server recvd EHLO from client-1 with payload: {u'role_name': u'develop'}
2018-01-02 17:09:47,559 [DEBUG] Got EHLO from client-1 with payload: {u'role_name': u'develop'}
2018-01-02 17:09:47,559 [DEBUG] Attempting to enroll client-1
2018-01-02 17:09:47,559 [DEBUG] Sending enroll to client-1
2018-01-02 17:09:47,567 [DEBUG] Starting new HTTP connection (1): localhost
2018-01-02 17:09:47,568 [DEBUG] http://localhost:8200 "GET /v1/auth/approle/role/develop/role-id HTTP/1.1" 200 208
2018-01-02 17:09:47,571 [DEBUG] http://localhost:8200 "POST /v1/auth/approle/role/develop/secret-id HTTP/1.1" 200 270
2018-01-02 17:09:47,574 [DEBUG] http://localhost:8200 "POST /v1/sys/wrapping/wrap HTTP/1.1" 200 318
2018-01-02 17:09:47,574 [DEBUG] server recvd EHLO from client-2 with payload: {u'role_name': u'develop'}
2018-01-02 17:09:47,574 [DEBUG] Got EHLO from client-2 with payload: {u'role_name': u'develop'}
2018-01-02 17:09:47,574 [DEBUG] Attempting to enroll client-2
2018-01-02 17:09:47,574 [DEBUG] Sending enroll to client-2
2018-01-02 17:09:47,575 [DEBUG] http://localhost:8200 "GET /v1/auth/approle/role/develop/role-id HTTP/1.1" 200 208
2018-01-02 17:09:47,577 [DEBUG] http://localhost:8200 "POST /v1/auth/approle/role/develop/secret-id HTTP/1.1" 200 270
2018-01-02 17:09:47,580 [DEBUG] http://localhost:8200 "POST /v1/sys/wrapping/wrap HTTP/1.1" 200 319
2018-01-02 17:09:47,581 [DEBUG] client-1 recvd ENROLL with payload: b95494d0-47c2-1c38-d50f-1b205164a6df
2018-01-02 17:09:47,581 [DEBUG] client-2 recvd ENROLL with payload: b344dfb4-35c9-d428-4817-22946c756dfc
2018-01-02 17:09:47,582 [DEBUG] client-2 handling enroll
2018-01-02 17:09:47,582 [DEBUG] client-1 handling enroll
2018-01-02 17:09:47,589 [DEBUG] Starting new HTTP connection (1): localhost
2018-01-02 17:09:47,589 [DEBUG] Starting new HTTP connection (1): localhost
2018-01-02 17:09:47,593 [DEBUG] http://localhost:8200 "POST /v1/sys/wrapping/unwrap HTTP/1.1" 200 258
2018-01-02 17:09:47,594 [DEBUG] client-2 unwrapped role-id: 8b352e16-81ff-bf72-5848-a5f2402d7fbb and secret-id: e5bf19ee-d1fe-b058-5391-364b3d6b6091
2018-01-02 17:09:47,594 [DEBUG] http://localhost:8200 "POST /v1/sys/wrapping/unwrap HTTP/1.1" 200 258
2018-01-02 17:09:47,594 [DEBUG] client-1 unwrapped role-id: 8b352e16-81ff-bf72-5848-a5f2402d7fbb and secret-id: f41ef4e3-e818-c09c-0774-91b118831782
2018-01-02 17:09:47,595 [DEBUG] Starting new HTTP connection (1): localhost
2018-01-02 17:09:47,596 [DEBUG] Starting new HTTP connection (1): localhost
2018-01-02 17:09:47,597 [DEBUG] http://localhost:8200 "POST /v1/auth/approle/login HTTP/1.1" 500 141
2018-01-02 17:09:47,598 [ERROR] client-1 Fatal exception in enroll_callback
2018-01-02 17:09:47,598 [DEBUG] http://localhost:8200 "POST /v1/auth/approle/login HTTP/1.1" 200 424
2018-01-02 17:09:47,598 [ERROR] alias "adf76e89-36ec-0087-761d-ff6a7a964287" in already tied to a different entity "6399d380-acc2-239b-1139-7cfd6d7ab439"
...
```

**Important Factoids:**

Occurs periodically in my test environment (two concurrent login client requests). I assume this is tied to network jitter on my loopback interface. Cannot be reproduced if a random sleep (up to 1s) is introduced between login requests.

Appears to be related with transaction alias sanity checking.

**References:**
<!-- Link to any references, such as GitHub issues or pull requests. -->

--- a/command/approle_concurrency_integ_test.go
+++ b/command/approle_concurrency_integ_test.go
@@ -1,0 +1,86 @@
+package command
+
+import (
+	"sync"
+	"testing"
+
+	"github.com/hashicorp/vault/api"
+	credAppRole "github.com/hashicorp/vault/builtin/credential/approle"
+	vaulthttp "github.com/hashicorp/vault/http"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/vault"
+	logxi "github.com/mgutz/logxi/v1"
+)
+
+func TestAppRole_Integ_ConcurrentLogins(t *testing.T) {
+	var err error
+	coreConfig := &vault.CoreConfig{
+		DisableMlock: true,
+		DisableCache: true,
+		Logger:       logxi.NullLog,
+		CredentialBackends: map[string]logical.Factory{
+			"approle": credAppRole.Factory,
+		},
+	}
+
+	cluster := vault.NewTestCluster(t, coreConfig, &vault.TestClusterOptions{
+		HandlerFunc: vaulthttp.Handler,
+	})
+
+	cluster.Start()
+	defer cluster.Cleanup()
+
+	cores := cluster.Cores
+
+	vault.TestWaitActive(t, cores[0].Core)
+
+	client := cores[0].Client
+
+	err = client.Sys().EnableAuthWithOptions("approle", &api.EnableAuthOptions{
+		Type: "approle",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = client.Logical().Write("auth/approle/role/role1", map[string]interface{}{
+		"bind_secret_id": "true",
+		"period":         "300",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	secret, err := client.Logical().Write("auth/approle/role/role1/secret-id", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	secretID := secret.Data["secret_id"].(string)
+
+	secret, err = client.Logical().Read("auth/approle/role/role1/role-id")
+	if err != nil {
+		t.Fatal(err)
+	}
+	roleID := secret.Data["role_id"].(string)
+
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < 100; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			secret, err = client.Logical().Write("auth/approle/login", map[string]interface{}{
+				"role_id":   roleID,
+				"secret_id": secretID,
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+			if secret.Auth.ClientToken == "" {
+				t.Fatalf("expected a successful login")
+			}
+		}()
+
+	}
+	wg.Wait()
+}

--- a/vault/identity_store.go
+++ b/vault/identity_store.go
@@ -281,9 +281,9 @@ func (i *IdentityStore) entityByAliasFactorsInTxn(txn *memdb.Txn, mountAccessor,
 	return i.MemDBEntityByAliasIDInTxn(txn, alias.ID, clone)
 }
 
-// CreateEntity creates a new entity. This is used by core to
+// CreateOrFetchEntity creates a new entity. This is used by core to
 // associate each login attempt by an alias to a unified entity in Vault.
-func (i *IdentityStore) FetchOrCreateEntity(alias *logical.Alias) (*identity.Entity, error) {
+func (i *IdentityStore) CreateOrFetchEntity(alias *logical.Alias) (*identity.Entity, error) {
 	var entity *identity.Entity
 	var err error
 

--- a/vault/identity_store.go
+++ b/vault/identity_store.go
@@ -249,7 +249,27 @@ func (i *IdentityStore) entityByAliasFactors(mountAccessor, aliasName string, cl
 		return nil, fmt.Errorf("missing alias name")
 	}
 
-	alias, err := i.MemDBAliasByFactors(mountAccessor, aliasName, false, false)
+	txn := i.db.Txn(false)
+
+	return i.entityByAliasFactorsInTxn(txn, mountAccessor, aliasName, clone)
+}
+
+// entityByAlaisFactorsInTxn fetches the entity based on factors of alias, i.e
+// mount accessor and the alias name.
+func (i *IdentityStore) entityByAliasFactorsInTxn(txn *memdb.Txn, mountAccessor, aliasName string, clone bool) (*identity.Entity, error) {
+	if txn == nil {
+		return nil, fmt.Errorf("nil txn")
+	}
+
+	if mountAccessor == "" {
+		return nil, fmt.Errorf("missing mount accessor")
+	}
+
+	if aliasName == "" {
+		return nil, fmt.Errorf("missing alias name")
+	}
+
+	alias, err := i.MemDBAliasByFactorsInTxn(txn, mountAccessor, aliasName, false, false)
 	if err != nil {
 		return nil, err
 	}
@@ -258,12 +278,12 @@ func (i *IdentityStore) entityByAliasFactors(mountAccessor, aliasName string, cl
 		return nil, nil
 	}
 
-	return i.MemDBEntityByAliasID(alias.ID, clone)
+	return i.MemDBEntityByAliasIDInTxn(txn, alias.ID, clone)
 }
 
 // CreateEntity creates a new entity. This is used by core to
 // associate each login attempt by an alias to a unified entity in Vault.
-func (i *IdentityStore) CreateEntity(alias *logical.Alias) (*identity.Entity, error) {
+func (i *IdentityStore) FetchOrCreateEntity(alias *logical.Alias) (*identity.Entity, error) {
 	var entity *identity.Entity
 	var err error
 
@@ -290,8 +310,23 @@ func (i *IdentityStore) CreateEntity(alias *logical.Alias) (*identity.Entity, er
 		return nil, err
 	}
 	if entity != nil {
-		return nil, fmt.Errorf("alias already belongs to a different entity")
+		return entity, nil
 	}
+
+	// Create a MemDB transaction to update both alias and entity
+	txn := i.db.Txn(true)
+	defer txn.Abort()
+
+	// Check if an entity was created before acquiring the lock
+	entity, err = i.entityByAliasFactorsInTxn(txn, alias.MountAccessor, alias.Name, false)
+	if err != nil {
+		return nil, err
+	}
+	if entity != nil {
+		return entity, nil
+	}
+
+	i.logger.Debug("identity: creating a new entity", "alias", alias)
 
 	entity = &identity.Entity{}
 
@@ -320,10 +355,12 @@ func (i *IdentityStore) CreateEntity(alias *logical.Alias) (*identity.Entity, er
 	}
 
 	// Update MemDB and persist entity object
-	err = i.upsertEntity(entity, nil, true)
+	err = i.upsertEntityInTxn(txn, entity, nil, true, false)
 	if err != nil {
 		return nil, err
 	}
+
+	txn.Commit()
 
 	return entity, nil
 }

--- a/vault/identity_store_test.go
+++ b/vault/identity_store_test.go
@@ -9,7 +9,7 @@ import (
 	"github.com/hashicorp/vault/logical"
 )
 
-func TestIdentityStore_CreateEntity(t *testing.T) {
+func TestIdentityStore_FetchOrCreateEntity(t *testing.T) {
 	is, ghAccessor, _ := testIdentityStoreWithGithubAuth(t)
 	alias := &logical.Alias{
 		MountType:     "github",
@@ -17,7 +17,7 @@ func TestIdentityStore_CreateEntity(t *testing.T) {
 		Name:          "githubuser",
 	}
 
-	entity, err := is.CreateEntity(alias)
+	entity, err := is.FetchOrCreateEntity(alias)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -33,10 +33,20 @@ func TestIdentityStore_CreateEntity(t *testing.T) {
 		t.Fatalf("bad: alias name; expected: %q, actual: %q", alias.Name, entity.Aliases[0].Name)
 	}
 
-	// Try recreating an entity with the same alias details. It should fail.
-	entity, err = is.CreateEntity(alias)
-	if err == nil {
-		t.Fatalf("expected an error")
+	entity, err = is.FetchOrCreateEntity(alias)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if entity == nil {
+		t.Fatalf("expected a non-nil entity")
+	}
+
+	if len(entity.Aliases) != 1 {
+		t.Fatalf("bad: length of aliases; expected: 1, actual: %d", len(entity.Aliases))
+	}
+
+	if entity.Aliases[0].Name != alias.Name {
+		t.Fatalf("bad: alias name; expected: %q, actual: %q", alias.Name, entity.Aliases[0].Name)
 	}
 }
 

--- a/vault/identity_store_test.go
+++ b/vault/identity_store_test.go
@@ -9,7 +9,7 @@ import (
 	"github.com/hashicorp/vault/logical"
 )
 
-func TestIdentityStore_FetchOrCreateEntity(t *testing.T) {
+func TestIdentityStore_CreateOrFetchEntity(t *testing.T) {
 	is, ghAccessor, _ := testIdentityStoreWithGithubAuth(t)
 	alias := &logical.Alias{
 		MountType:     "github",
@@ -17,7 +17,7 @@ func TestIdentityStore_FetchOrCreateEntity(t *testing.T) {
 		Name:          "githubuser",
 	}
 
-	entity, err := is.FetchOrCreateEntity(alias)
+	entity, err := is.CreateOrFetchEntity(alias)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -33,7 +33,7 @@ func TestIdentityStore_FetchOrCreateEntity(t *testing.T) {
 		t.Fatalf("bad: alias name; expected: %q, actual: %q", alias.Name, entity.Aliases[0].Name)
 	}
 
-	entity, err = is.FetchOrCreateEntity(alias)
+	entity, err = is.CreateOrFetchEntity(alias)
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/vault/identity_store_util.go
+++ b/vault/identity_store_util.go
@@ -666,12 +666,29 @@ func (i *IdentityStore) MemDBAliasByFactors(mountAccessor, aliasName string, clo
 		return nil, fmt.Errorf("missing mount accessor")
 	}
 
+	txn := i.db.Txn(false)
+
+	return i.MemDBAliasByFactorsInTxn(txn, mountAccessor, aliasName, clone, groupAlias)
+}
+
+func (i *IdentityStore) MemDBAliasByFactorsInTxn(txn *memdb.Txn, mountAccessor, aliasName string, clone bool, groupAlias bool) (*identity.Alias, error) {
+	if txn == nil {
+		return nil, fmt.Errorf("nil txn")
+	}
+
+	if aliasName == "" {
+		return nil, fmt.Errorf("missing alias name")
+	}
+
+	if mountAccessor == "" {
+		return nil, fmt.Errorf("missing mount accessor")
+	}
+
 	tableName := entityAliasesTable
 	if groupAlias {
 		tableName = groupAliasesTable
 	}
 
-	txn := i.db.Txn(false)
 	aliasRaw, err := txn.First(tableName, "factors", mountAccessor, aliasName)
 	if err != nil {
 		return nil, fmt.Errorf("failed to fetch alias from memdb using factors: %v", err)

--- a/vault/request_handling.go
+++ b/vault/request_handling.go
@@ -438,7 +438,7 @@ func (c *Core) handleLoginRequest(ctx context.Context, req *logical.Request) (re
 
 			// Fetch the entity for the alias, or create an entity if one
 			// doesn't exist.
-			entity, err = c.identityStore.FetchOrCreateEntity(auth.Alias)
+			entity, err = c.identityStore.CreateOrFetchEntity(auth.Alias)
 			if err != nil {
 				return nil, nil, err
 			}

--- a/vault/request_handling.go
+++ b/vault/request_handling.go
@@ -436,22 +436,15 @@ func (c *Core) handleLoginRequest(ctx context.Context, req *logical.Request) (re
 
 			var err error
 
-			// Check if an entity already exists for the given alias
-			entity, err = c.identityStore.entityByAliasFactors(auth.Alias.MountAccessor, auth.Alias.Name, false)
+			// Fetch the entity for the alias, or create an entity if one
+			// doesn't exist.
+			entity, err = c.identityStore.FetchOrCreateEntity(auth.Alias)
 			if err != nil {
 				return nil, nil, err
 			}
 
-			// If not, create one.
 			if entity == nil {
-				c.logger.Debug("core: creating a new entity", "alias", auth.Alias)
-				entity, err = c.identityStore.CreateEntity(auth.Alias)
-				if err != nil {
-					return nil, nil, err
-				}
-				if entity == nil {
-					return nil, nil, fmt.Errorf("failed to create an entity for the authenticated alias")
-				}
+				return nil, nil, fmt.Errorf("failed to create an entity for the authenticated alias")
 			}
 
 			auth.EntityID = entity.ID

